modules/masterbar/masterbar.php

```diff
@@ -68,6 +118,50 @@ function __construct() {
 		add_action( 'wp_logout', array( $this, 'maybe_logout_user_from_wpcom' ) );
 	}
 
+	public function tracks_and_redirect_add_route() {
+		add_rewrite_rule( 'jetpack-track-and-bounce.php\?tracks_and_bounce=(\w*)', 'index.php?tracks_and_bounce=$matches[1]', 'top' );
+	}
+
+	public function tracks_and_redirect_add_query_vars( $query_vars ) {
+		$query_vars[] = 'tracks_and_bounce';
+		$query_vars[] = 'tracks_and_bounce_nonce';
+		return $query_vars;
+	}
+
+	public function tracks_and_redirect_parse_request( &$wp ) {
+		if (
+			array_key_exists( 'tracks_and_bounce', $wp->query_vars ) &&
+			array_key_exists( $_GET[ 'tracks_and_bounce' ], $this->links_tracks_events ) &&
+			wp_verify_nonce( $_GET[ 'tracks_and_bounce_nonce' ], 'jp-tracks-masterbar-nonce' )
+		) {
+			$target = $this->tracks_and_redirect_get_target( $_GET[ 'tracks_and_bounce' ] );
+
+			JetpackTracking::record_user_event(
+				$this->links_tracks_events[ $_GET[ 'tracks_and_bounce' ] ][0],
+				$data = array( 'target' => $target )
+			);
+			wp_redirect( $target );
+			die();
+		}
+
+		JetpackTracking::record_user_event(
+			'masterbar_click_error',
+			$data = array( 'target' => $this->tracks_and_redirect_get_target( $_GET[ 'tracks_and_bounce' ] ) )
+		);
+		// no nonce, wrong link or missing tracking information, push back to settings page
+		wp_safe_redirect(
+			add_query_arg(
+				array( 'page' => 'jetpack-settings' ),
+				network_admin_url( 'admin.php' )
+			)
+		);
+		die();
+	}
+
+	private function tracks_and_redirect_get_target( $trackId ) {
+		return "http://wordpress.com" . str_replace("{blog_url}", Jetpack_Options::get_option( 'id' ),  $this->links_tracks_events[ $trackId ][1] );
```

Comment:
I think we should be able to use single quotes here around `http://wordpress.com` and `{blog_url}`. Also, maybe consider using `https` instead of `http` in the wordpress.com url.

Reply:
Agreed, we should use `https` instead. Also we can replace `Jetpack_Options::get_option( 'id' )` with the existing `$this->primary_site_url`.

Also would it make sense to simplify this and avoid `str_replace` completely. We can initialize `links_tracks_events` in constructor after `primary_site_url` is set, so we don't have to use `{blog_url}` placeholder.

Reply:
Sorry for the typo. There's no emoji for shame...

I've updated the code to initialize the tracks events on the constructor and remove the `tracks_and_redirect_get_target ` altogether on a36123157b1f1350f06c11dc27d73f714c68a981.